### PR TITLE
Added taskstack builder return intent

### PR DESCRIPTION
**Pull request by dudego**
#63 task back stack (Also added a method in sample with params)

You can also annotate static methods that take a `Context` and return a `TaskStackBuilder`. `DeepLinkDispatch` will call that
method to create `Intent` from `TaskStackBuilder` last `Intent` and use it when starting your Activity via that registered deep link:

```java
@DeepLink("http://example.com/deepLink/{id}/{name}")
  public static TaskStackBuilder intentForTaskStackBuilderMethods(Context context) {
    Intent detailsIntent =  new Intent(context, SecondActivity.class).setAction(ACTION_DEEP_LINK_COMPLEX);
    Intent parentIntent =  new Intent(context, MainActivity.class).setAction(ACTION_DEEP_LINK_COMPLEX);
    TaskStackBuilder  taskStackBuilder = TaskStackBuilder.create(context);
    taskStackBuilder.addNextIntent(parentIntent);
    taskStackBuilder.addNextIntent(detailsIntent);
    return taskStackBuilder;

  }
```

### CODE REVIEW

File: deeplinkdispatch-processor/build.gradle [hunk 1]
```diff
@@ -13,7 +13,8 @@ dependencies {
     testCompile 'junit:junit:4.12'
     testCompile 'org.assertj:assertj-core:1.7.0'
     testCompile 'com.google.android:android:4.1.1.4'
-    testCompile 'com.google.android:support-v4:r7'
+    //testCompile 'com.google.android:support-v4:r7'
+    testCompile fileTree(dir: 'libs', include: ['*.jar'])
```
Comment (felipecsl) [inline deeplinkdispatch-processor/build.gradle]: why?

**Comment by felipecsl** [comment]
Seems like you should fix your formatting before we can review this

File: deeplinkdispatch-processor/build.gradle (hunk 1 above)
Comment (dudego) [inline deeplinkdispatch-processor/build.gradle]: Because `taskStackBuilder.editIntentAt(int index)` is not present in v4:r7.

File: deeplinkdispatch-processor/build.gradle [hunk 2]
```diff
@@ -13,7 +13,8 @@ dependencies {
     testCompile 'junit:junit:4.12'
     testCompile 'org.assertj:assertj-core:1.7.0'
     testCompile 'com.google.android:android:4.1.1.4'
-    testCompile 'com.google.android:support-v4:r7'
+    //testCompile 'com.google.android:support-v4:r7'
+    testCompile fileTree(dir: 'libs', include: ['android-support-v4.jar'])
```
Comment (felipecsl) [inline deeplinkdispatch-processor/build.gradle]: please revert this change

File: README.md [hunk 3]
```diff
@@ -62,6 +62,22 @@ public static Intent intentForDeepLinkMethod(Context context) {
 }
 ```
 
+
+You can also annotate static methods that take a `Context` and return a `TaskStackBuilder`. `DeepLinkDispatch` will call that
+method to create `Intent` from `TaskStackBuilder` last `Intent` and use it when starting your Activity via that registered deep link:
+
+```java
+@DeepLink("http://example.com/deepLink/{id}/{name}")
+  public static TaskStackBuilder intentForTaskStackBuilderMethods(Context context) {
```
Comment (felipecsl) [inline README.md]: `TaskStackBuilder` requires API 16 and newer. You should mention that in the README too

File: README.md [hunk 4]
```diff
@@ -62,6 +62,22 @@ public static Intent intentForDeepLinkMethod(Context context) {
 }
 ```
 
+
+You can also annotate static methods that take a `Context` and return a `TaskStackBuilder`. `DeepLinkDispatch` will call that
+method to create `Intent` from `TaskStackBuilder` last `Intent` and use it when starting your Activity via that registered deep link:
+
+```java
+@DeepLink("http://example.com/deepLink/{id}/{name}")
```
Comment (felipecsl) [inline README.md]: nit: formatting is off on this line

File: README.md [hunk 5]
```diff
@@ -62,6 +62,22 @@ public static Intent intentForDeepLinkMethod(Context context) {
...
+  public static TaskStackBuilder intentForTaskStackBuilderMethods(Context context) {
+    Intent detailsIntent =  new Intent(context, SecondActivity.class).setAction(ACTION_DEEP_LINK_COMPLEX);
+    Intent parentIntent =  new Intent(context, MainActivity.class).setAction(ACTION_DEEP_LINK_COMPLEX);
+    TaskStackBuilder  taskStackBuilder = TaskStackBuilder.create(context);
+    taskStackBuilder.addNextIntent(parentIntent);
+    taskStackBuilder.addNextIntent(detailsIntent);
+    return taskStackBuilder;
+
```
Comment (felipecsl) [inline README.md]: remove blank line

File: README.md [hunk 6]
```diff
@@ -62,6 +62,22 @@ public static Intent intentForDeepLinkMethod(Context context) {
...
 
+
+You can also annotate static methods that take a `Context` and return a `TaskStackBuilder`. `DeepLinkDispatch` will call that
+method to create `Intent` from `TaskStackBuilder` last `Intent` and use it when starting your Activity via that registered deep link:
+
+```java
+@DeepLink("http://example.com/deepLink/{id}/{name}")
+  public static TaskStackBuilder intentForTaskStackBuilderMethods(Context context) {
+    Intent detailsIntent =  new Intent(context, SecondActivity.class).setAction(ACTION_DEEP_LINK_COMPLEX);
+    Intent parentIntent =  new Intent(context, MainActivity.class).setAction(ACTION_DEEP_LINK_COMPLEX);
```
Comment (felipecsl) [inline README.md]: double spaces

File: README.md (hunk 3 above)
Comment (felipecsl) [inline README.md]: Oh, never mind, this is using the support-v4 version

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java [hunk 7]
```diff
@@ -371,20 +371,45 @@ private void generateDeepLinkDelegate() throws IOException {
...
         .addStatement("$T method", Method.class)
         .beginControlFlow("try")
         .addStatement("method = c.getMethod(entry.getMethod(), $T.class)",
                 ClassName.get("android.content", "Context"))
+        .beginControlFlow("if (method.getReturnType().equals($T.class))",
+                ClassName.get("android.support.v4.app", "TaskStackBuilder"))
+        .addStatement("taskStackBuilder = (TaskStackBuilder) method.invoke(c, activity)")
+        .beginControlFlow("if (taskStackBuilder.getIntentCount() == 0)")
+        .addStatement("return createResultAndNotify(activity, false, uri, \"Could not deep "
+                + "link to method: \" + entry.getMethod() + \" intents length ==0\" )")
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.java]: formatting `==0`

**Comment by felipecsl** [comment]
this needs unit tests for the new functionality

File: deeplinkdispatch-processor/build.gradle (hunk 2 above)
Comment (dudego) [inline deeplinkdispatch-processor/build.gradle]: if i revert this then DeepLinkProcessor tests fail to compile as v4:r7 doesn't contain `TaskStackBuillder.editIntentAt(int)` method  but  ['getIntent(int index)'](https://developer.android.com/reference/android/support/v4/app/TaskStackBuilder.html#getIntent(int))  which is deprecated now

Comment (felipecsl) [inline deeplinkdispatch-processor/build.gradle]: you should be able to use this instead, since it's available on jcenter:
```
testCompile 'com.google.android:support-v4:24.1.1'
```

Comment (dudego) [inline deeplinkdispatch-processor/build.gradle]: i've tried this but no success. Dependency is resolved on syncing the project but not at run time.
jcenter provides aar with contains jar inside libs for android.support.v4.TaskStackBuilder which can't be resolved at runtime...
This is similar to issue referred in [ Jar files in libs folder of in AAR package was not picked up](https://github.com/simpligility/android-maven-plugin/issues/484)

**Comment by felipecsl** [comment]
thanks!
